### ISSUE TITLE
GPMSA with shifted non [0,1] prior

### ISSUE THREAD

**Issue by briadam** [libqueso/queso#575]
Per discussion with @roystgnr and @dmcdougall I believe the attached variant of test_gpmsa_vector.C reproduces the issue I'm seeing from Dakota.  The calibration domain appears wrong in the aggregate GPMSA prior, and something downstream breaks.  Yet the distributionMean gives the right initial point...

I expect the calibration to be performed over the [-1, 0] hypercube for the calibration parameters.  What I see instead is (truncated):

```
BMA: priorRv PDF domainSet on [-1, 0]
In BoxSubset<V,M>::print(): m_minValues = -1 -1 -1 -1 -1 , m_maxValues = 0 0 0 0 0 , m_volume = 1

BMA: GPMSA prior PDF domainSet (calib params now on [0,1]!?!
In BoxSubset<V,M>::print(): m_minValues = 0 0 0 0 0 0.3 0.3 0.3 0 0 0 0 0 0 0 0 60 0.3 , m_maxValues = 1 1 1 1 1 inf inf inf 1 1 1 1 1 1 inf 1 100000 inf , m_volume = inf
```

Yet somehow the distributionMean is still on [-1.0, 0.0].  And despite the initial values appearing to be between those bounds...

```
BMA: Default initials
-0.5 -0.5 -0.5 -0.5 -0.5 1 1 1 0.909091 0.909091 0.909091 0.909091 0.909091 0.909091 10000 0.909091 999.999 1 
BMA: Final initials
-0.5 -0.5 -0.5 -0.5 -0.5 0.4 0.4 0.4 0.97 0.97 0.97 0.97 0.2 0.8 10 0.97 8000 1

Assertion `!(outOfTargetSupport)' failed.
initial position should not be out of target pdf support
../../src/stats/src/MetropolisHastingsSG.C, line 1405, compiled Apr 19 2017 at 09:29:14
```

[test_gpmsa_vector.C.txt](https://github.com/libqueso/queso/files/953107/test_gpmsa_vector.C.txt)

**Comment by briadam** [comment]
Note that my example doesn't shift the simulation data to [-1, 0], so the example isn't complete.  I was just trying to demonstrate what might be an issue with the prior...

This might be causing this issue: https://github.com/libqueso/queso/blob/33ab1ef36c3a2776a48fb01cade6b1be40c635d1/src/gp/src/GPMSA.C#L1426  Sets the totalMin/Max to 0/1, yet the bounds are never pulled off the m_parameterPrior or other object, so they are left at [0,1]...

**Comment by briadam** [comment]
I confirmed that setting the first dimParameter entries of totalMins/totalMaxs in GPMSAFactory<V, M>::setUpHyperpriors() to the actual bounds for my problem gets around this issue.  However, m_parameterSpace and m_parameterPrior don't necessarily have enough information to set bounds (unless cast to an object that does), so I don't have an idea for a better solution...

I am glad we have a variable called dimSum; though it makes me hungry... ;-)

**Comment by roystgnr** [comment]
It looks like the long term fix here is to add min/max getters to VectorSet.h?

Short-term... when we query from the total prior concatenatedRV it will pass that to the parameter priors, right?  So what are the total prior mins and maxes being used for?  Can we pass -infinity, infinity here?

**Comment by dmcdougall** [comment]
> It looks like the long term fix here is to add min/max getters to VectorSet.h?

#557

**Comment by roystgnr** [comment]
Where by "long term" I apparently mean "two weeks ago"...

**Comment by briadam** [comment]
Just a note that setting min/max to -/+ inf (as I too had attempted as a workaround) is not helpful.  With a bounded uniform prior, I end up getting posterior samples well outside the bounds (order of magnitude).  I haven't thought through whether this is surprising or expected, but for now, I'm hard-wiring my prior bounds as the min/max.

**Comment by briadam** [comment]
FWIW, if you think of a compact workaround for this, or decide to actually pull some of the changes from #557 I'd consider doing a local patch to our copy of QUESO.  I don't see an easy client side workaround due to the wide array of distributions we support.  We have a variable transform that can (nonlinearly) map all the distros to [-1, 1], but it's not easy to reconfigure it to [0, 1] right now.

**Comment by briadam** [comment]
FWIW, if you think of a compact workaround for this, or decide to actually pull some of the changes from #557 I'd consider doing a local patch to our copy of QUESO.  I don't see an easy client side workaround due to the wide array of distributions we support.  We have a variable transform that can (nonlinearly) map all the distros to [-1, 1], but it's not easy to reconfigure it to [0, 1] right now.

**Comment by roystgnr** [comment]
Wait a minute.  We don't want to query mins/maxes for the parameter prior, we *really* want a tensor product of the domain sets, right?  So we ought to be querying the prior.imageSet(), then doing a ConcatenationSubset of that with the BoxSubset we construct for the hyperparameters?

Let me try setting up a PR with that for you to play with.

**Comment by briadam** [comment]
Yes, I think that makes sense.  I guess I was too prescriptive in suggesting a solution and didn't realize the prior carried along with it its domain in that way.  Sounds good.  This also makes me realize that our current Dakota interface to QUESO is always setting the parameter domain to the bounds of the prior and the user might not always (ever?) want that...

**Comment by roystgnr** [comment]
Damn.  Breaks at

```
    transformInitialCovMatrixToGaussianSpace(
        dynamic_cast<const BoxSubset<P_V, P_M> & >(m_targetPdf.domainSet()));
```

because the logit transform needs mins and maxes and BoxSubset is the only thing that (currently) provides them.  So we really will need to wait for #557 to fix this.

**Comment by dmcdougall** [comment]
You can turn off the logit transform:

```
ip_mh_tk = random_walk
ip_mh_algorithm = random_walk
```

**Comment by roystgnr** [comment]
"Doctor, it hurts when I lift my arm!"
"So stop lifting your arm!"

**Comment by dmcdougall** [comment]
Doctor of Philosophy, not Medical Doctor.  I accept no responsibility.

That said, #557 should fix your problem but it won't make it into a bugfix release.  How do you want to proceed?

**Comment by briadam** [comment]
I agree that the scope of #557 seems too great for a patch release.  For purposes of Dakota 6.6, stability of QUESO DRAM is more crucial than rushing more GPMSA completeness, particularly before we get better verification testing in place (I should be able to contribute Brian's "linear verification" problem back to QUESO examples/ soon and hope we can eventually promote it to a test).  There are still a lot of unresolved questions with GPMSA, even for the scalar and multivariate cases, and we'll need complete, numerically robust, scaling, better options parsing, and good tests before we make this a generally available capability.

I'd propose that I just stabilize what I have on the Dakota side (which already includes the patches from #571 and #573), and let @brianw525 experiment with what's there.  For now, I'll augment my Dakota GPMSAOptions to pass theta bounds and extract them as needed in GPMSA when setting up the concatenated priors.  This way, Roy can proceed with leveraging #557 without the pressure of an immediate deadline, and I can wait for the next major release to get at that...

**Comment by dmcdougall** [comment]
Fixed in #577.  Please re-open @briadam if that isn't the case.
